Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,354 @@
+  private let minimumSentPingIntervalWithoutData: TimeAmount
+
+  /// If there are no data/header frames being sent:
+  /// The minimum amount of time expected between receiving successive pings.
+  /// If the time between successive pings is less than this value, then the ping will be considered a bad ping from the peer.
+  /// Such a ping counts as a "ping strike".
+  /// Ping strikes are only applicable to server handler
+  private let minimumReceivedPingIntervalWithoutData: TimeAmount?
+
+  /// Maximum number of bad pings that the server will tolerate before sending an HTTP2 GOAWAY frame and closing the connection.
+  /// Setting it to `0` allows the server to accept any number of bad pings.
+  /// Ping strikes are only applicable to server handler
+  private let maximumPingStrikes: UInt?
+
+  /// When the last ping was received
+  private var lastReceivedPingDate: NIODeadline? = nil
+
+  /// When the last ping was sent
+  private var lastSentPingDate: NIODeadline? = nil
+
+  /// The number of pings sent on the transport without any data
+  private var sentPingsWithoutData = 0
+
+  /// Number of strikes
+  private var pingStrikes: UInt = 0
+
+  /// The scheduled task which will close the connection.
+  private var scheduledClose: Scheduled<Void>? = nil
+
+  /// Number of active streams
+  private var activeStreams = 0 {
+    didSet {
+      if activeStreams > 0 {
+        self.sentPingsWithoutData = 0
+      }
+    }
+  }
+
+  private let goAwayFrame = HTTP2Frame.FramePayload.goAway(lastStreamID: .rootStream, errorCode: .enhanceYourCalm, opaqueData: nil)
+
+  // For testing only
+  var testingOnlyNow: NIODeadline?
+
+  enum Action {
+    case none
+    case schedulePing(delay: TimeAmount, timeout: TimeAmount)
+    case cancelScheduledPing
+    case cancelScheduledTimeout
+    case reply(HTTP2Frame.FramePayload)
+  }
+
+  init(
+    pingCode: UInt64,
+    interval: TimeAmount,
+    timeout: TimeAmount,
+    permitWithoutCalls: Bool,
+    maximumPingsWithoutData: UInt,
+    minimumSentPingIntervalWithoutData: TimeAmount,
+    minimumReceivedPingIntervalWithoutData: TimeAmount? = nil,
+    maximumPingStrikes: UInt? = nil
+  ) {
+    self.pingCode = pingCode
+    self.interval = interval
+    self.timeout = timeout
+    self.permitWithoutCalls = permitWithoutCalls
+    self.maximumPingsWithoutData = maximumPingsWithoutData
+    self.minimumSentPingIntervalWithoutData = minimumSentPingIntervalWithoutData
+    self.minimumReceivedPingIntervalWithoutData = minimumReceivedPingIntervalWithoutData
+    self.maximumPingStrikes = maximumPingStrikes
+  }
+
+  func streamCreated() -> Action {
+    self.activeStreams += 1
+
+    if self.activeStreams == 1 {
+      return .schedulePing(delay: self.interval, timeout: self.timeout)
+    } else {
+      return .none
+    }
+  }
+
+  func streamClosed() -> Action {
+    self.activeStreams -= 1
+
+    if self.activeStreams == 0 && !self.permitWithoutCalls {
+      return .cancelScheduledPing
```

Comment:
This diverges slightly from the implied documented behaviour of grpc, which is that the keep alive timer fires cyclically, once per `GRPC_ARG_KEEPALIVE_TIME_MS` from connection completion. The ping is only _emitted_ based on some filtering criteria, but the timer still fires on this regular interval.

Given that you have this filtering criteria in place anyway, should we use a repeating call instead of a one-off scheduled call, and just cancel the repeating call in handlerRemoved?

Reply:
You're right, I already check this in `shouldBlockPing` so no need to add a check here.